### PR TITLE
Email verification

### PR DESCRIPTION
**Pull request by MichelBodde**
## Purpose
The purpose of this branch is to add the functionality of email verification to our log-in and sign-up process to minimize low-effort fake-accounts.

## Outcome
If the code functions as intended, new users registering for an account will receive an email after succesfully entering their data in the sign-up form. In this email they will find a link needed for further registration.
Furthermore, log-in is now also gated behind verification.

## Details
SQL changes: in the table "gebruiker", 2 columns were added to support this new verification, 'is_geverifieerd' and 'vkey'. An alternative to this implementation would have been to use session variables to temporarily store all sign-up data but I personally think that is not secure enough.

The Controllers have been changed in several ways: 
- Firstly, the input sanitisation has been standardized under the secure_input function from the Core/Controlller file. 
- Secondly, when a user makes encounters an error whilst filling out a form, filled in textfields are carried over upon reaching header("Location: ...").
- Thirdly, the verification system is added. After signup, the username and current timestamp is saved as that user's verificationkey. An email is sent to the adres the user provided in the form containing a link to the login page with the vkey as get-variable attached. When that key matches the one in the dtb, the user is verified and allowed to log in.

Views have been updated with aditional success and error alerts and the functionality to check vkey's after signup.

Models have received new functions and updated existing functions.

### CODE REVIEW

**Comment by Stijnwingens** [comment]
In your PR at the end of section Purpose you mention "login is now also gated behind verification." What exactly do you mean?

**Comment by MichelBodde** [comment]
> 
> 
> In your PR at the end of section Purpose you mention "login is now also gated behind verification." What exactly do you mean?

That users can not log in unless they have already been verified. More specifically "is_geverifieerd" needs to be set to 1

File: controller/registrerenController.php [hunk 1]
```diff
@@ -30,68 +30,141 @@ function index() {
...
-            $antwoordtekst = strip_tags((isset($_POST['antwoordtekst']) ? $_POST['antwoordtekst'] : null));
+            $gebruikersnaam = secure_input((isset($_POST['uid']) ? $_POST['uid'] : null));
+            $voornaam = secure_input((isset($_POST['voornaam']) ? $_POST['voornaam'] : null));
+            $tussenvoegsel = secure_input((isset($_POST['tussenvoegsel']) ? $_POST['tussenvoegsel'] : null));
+            $achternaam = secure_input((isset($_POST['achternaam']) ? $_POST['achternaam'] : null));
+            $adresregel_1 = secure_input((isset($_POST['adres_1']) ? $_POST['adres_1'] : null));
+            $adresregel_2 = secure_input((isset($_POST['adres_2']) ? $_POST['adres_2'] : null));
+            $postcode = secure_input((isset($_POST['postcode']) ? $_POST['postcode'] : null));
+            $plaatsnaam = secure_input((isset($_POST['plaatsnaam']) ? $_POST['plaatsnaam'] : null));
+            $land_id = secure_input((isset($_POST['land_id']) ? $_POST['land_id'] : null));
```
Comment (pepijnwinkelmolen) [inline controller/registrerenController.php]: hadden we deze nu wel weer nodig? dacht dat die verwijdert was omdat het altijd 6030 zou zijn

**Comment by Stijnwingens** [comment]
> 
> 
> > In your PR at the end of section Purpose you mention "login is now also gated behind verification." What exactly do you mean?
> 
> That users can not log in unless they have already been verified. More specifically "is_geverifieerd" needs to be set to 1

A like that. Great!

File: controller/loginController.php [hunk 2]
```diff
@@ -4,28 +4,35 @@ class loginController extends Controller {
...
-                } else {
-                    $mailuid = strip_tags((isset($_POST['gebruikersnaam']) ? $_POST['gebruikersnaam'] : null));
+                    $mailuid = secure_input((isset($_POST['gebruikersnaam']) ? $_POST['gebruikersnaam'] : null));
 
-                    $password = strip_tags((isset($_POST['wachtwoord']) ? $_POST['wachtwoord'] : null));
+                    $password = secure_input((isset($_POST['wachtwoord']) ? $_POST['wachtwoord'] : null));
 
                     if (empty($mailuid) || empty($password)) {
                         $data['error_input'] = "empty_fields";
+                        header("Location: ../login?error=emptyfields&gebruikersnaam=".$mailuid);
```
Comment (pepijnwinkelmolen) [inline controller/loginController.php]: ik zie persoonlijk niet graag variabelen teruggegeven zelfs al zou dit niet veel kwaad kunnen

File: DB SQL/03 - CREATE gebruiker.sql [hunk 3]
```diff
@@ -18,6 +18,8 @@ CREATE TABLE gebruiker (
 	vraag INT NOT NULL,
 	antwoordtekst VARCHAR(20) NOT NULL,
 	rating NUMERIC(4,1) NOT NULL,
+	is_geverifieerd BIT DEFAULT 0,
```
[the remark this answers is not in the document]
Comment (MichelBodde) [reply DB SQL/03 - CREATE gebruiker.sql]: Is_geverifieerd is not a field for entry that users can see or interact with. It's purpose is to keep track of whether or not the user has reached the loginpage with a valid verificationkey. By default(upon signup/new entry in dtb) users are not verified, thus 0.

File: DB SQL/03 - CREATE gebruiker.sql [hunk 4]
```diff
@@ -18,6 +18,8 @@ CREATE TABLE gebruiker (
...
+	vkey CHAR(255) NOT NULL,
```
[the remark this answers is not in the document]
Comment (MichelBodde) [reply DB SQL/03 - CREATE gebruiker.sql]: done

File: model/userAuthModel.php [hunk 5]
```diff
@@ -1,19 +0,0 @@
-<?php
```
[the remark this answers is not in the document]
Comment (MichelBodde) [reply model/userAuthModel.php]: deze hele file is weggehaald omdat er duplicate functions in zaten. Min of meer dezelfde twee functies die in loginModel en registratieModel  staan, maar dan outdated.

File: view/userregistreren.php [hunk 6]
```diff
@@ -1,6 +1,11 @@
 <h1 class="uk-margin-top" style="margin-left: calc(50% - 99px);">Registratie</h1>
 <?php
-if (isset($this->vars['error_input'])) {
+if (isset($_GET["registration"])) {
```
[the remark this answers is not in the document]
Comment (MichelBodde) [reply view/userregistreren.php]: Ik weet persoonlijk niet of dat makkelijker kan, maar ik heb als standaard format voor get altijd een isset bij een nieuwe get-naam zodat als die naam er niet in voorkomt een hele lap code kan worden overgeslagen.
In dit specifieke geval heeft het niet veel effect.

File: controller/registrerenController.php (hunk 1 above)
Comment (MichelBodde) [reply controller/registrerenController.php]: Niet persé maar heb het er wel ingehouden om de code consistent te houden.

File: view/userregistreren.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (Stijnwingens) [reply view/userregistreren.php]: Er expliciet niet vanuit gaan dat een variabele is geset is een goede en belangrijke check om te zorgen dat hackers minder informatie krijgen in de vorm van errors. Dit afvangen is zeker aanbevolen om te gebruiken, of het nu nodig is of niet. Het liefst altijd doen dus.

File: controller/registrerenController.php (hunk 1 above)
Comment (Stijnwingens) [reply controller/registrerenController.php]: Same as my other comment earlier. Good work Michel!

File: view/includes/registratieformulier.inc.php [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                    </div>
+                </div>
+            </div>
+            <div class="uk-width-1-1 uk-margin-top"> <!--uk-width-1-2@s-->
+                <label class="uk-form-label" for="vraag">Beveiligingsvraag</label>
+                <div class="uk-form-controls">
+                    <div class="uk-inline uk-width-1-1">
+                        <select class="uk-select" name="vraag" id="vraag" value="<?= (isset($_POST['vraag']) ? $_POST['vraag'] : null); ?>">
+                            <option selected disabled>Kies een vraag</option>
+                            <?php echo $this->vars['vragen']; ?>
```
Comment (IvanTeunesen) [inline view/includes/registratieformulier.inc.php]: Could be changed to `<?= $this->vars['vragen]; ?>`, but not necessary.

**Pull request merged**